File: bigbear/616/129.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int sum;
    public int sumNumbers(TreeNode root) {
        sum = 0;
        add(root, 0);
        return sum;
    }
    
    private void add(TreeNode root, int subsum){
        if(root == null){
            return;
        }
        subsum = subsum * 10 + root.val;
        
        if(root.left ==null && root.right == null) {
            sum += subsum;
        }
        
        add(root.left, subsum);
        add(root.right, subsum);
    }
}

File: jenfengchang/129.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    int res = 0;
public:
    int sumNumbers(TreeNode* root) {
        InOrderTraversal(root, 0);
        return res;
    }
    void InOrderTraversal(TreeNode* root, int sum) {
        if(root == NULL)
            return;
        sum = sum * 10 + root->val;
        if(root->left == NULL & root->right == NULL) {
            res += sum;
        }
        InOrderTraversal(root->left, sum);
        InOrderTraversal(root->right, sum);
    }
};
